compiler/rustc_middle/src/ty/util.rs

```diff
@@ -828,6 +832,61 @@ impl<'tcx> ty::TyS<'tcx> {
         }
     }
 
+    pub fn gc_layout(&'tcx self, tcx: TyCtxt<'tcx>, param_env: ty::ParamEnv<'tcx>) -> (u64, u64) {
+        let layout = tcx.layout_of(param_env.and(self)).unwrap();
+        let w_size = Size::from_bytes(size_of::<usize>());
+        let w_align = Align::from_bytes(w_size.bytes()).unwrap();
+
+        if layout.align.abi.bytes() < (w_size.bytes())
+            || self.is_no_trace(tcx.at(DUMMY_SP), param_env)
+        {
+            // If an ADT is aligned to a size smaller than 1 word, then it
```

Comment:
I'm not sure if this is safe: I don't think that Rust's definition of alignment forces us to store word sized things on a word boundary. Now, admittedly, Boehm presumably only checks at a word sized boundary, so this optimisation is probably safe in the sense that it won't make the situation any worse than is already the case in Boehm. However -- assuming I've not got everything backwards! -- it might be good to be clear about this.

Reply:
This comment is hinting at something subtly different, what it's trying to say is that a struct (ADT) in Rust will be aligned to multiples of the largest field. If that alignment is less than 1 word, then we know that the struct doesn't contain any fields large enough to store a pointer. 

However, you do raise a good point, which is more generally related to packed structs. In fact, as you say, even Boehm won't be able to detect pointers in packed structs. This is also a problem in cppgc. I definitely agree that we should make it clear to the user that using packed structs is here-be-dragons territory and that one shouldn't expect any pointers inside to be traced.

Reply:
I should clarify, packed structs are not inherently dangerous, but packed structs which contain pointers on non-word boundaries.

Reply:
Can we detect packed structs and bork on them or output a warning perhaps?

Reply:
Yes, we definitely could. I'm not sure this is the right place to do it though, as it's also a problem for regular conservative scanning. How about a new PR which borks on any packed struct which has a potential pointer field at a non-word offset?

Reply:
I'm a bit worried that if we leave it we'll forget. The problem is that a user might not know they're relying on a packed struct to point to them, so it's a bit more dangerous than it first seems. I'd be fine with us just having a `todo` for the time being (assuming that it doesn't stop rustgc from building!).

Reply:
This code-path is only invoked when the type is being allocated precisely though. So even adding a `todo` here will miss ADTs which are scanned conservatively. I can raise a PR in parallel, and have this rebased against it, or I can add this functionality in a separate commit. I prefer the former tbh, because it's an orthogonal change.

Reply:
Raising a parallel PR works for me!